File: examples/oxidationFrontPS/oxidationFrontPS.py
import sys
import os
import viennaps as vps
import numpy as np
from scipy.sparse import coo_matrix

import geometryPS

# Material IDs
MAT_SUBSTRATE = 0
MAT_OXIDE = 1
MAT_MASK = 2
MAT_AMBIENT = 3

class OxidationSimulation:
    """
    Electric Field Driven Oxidation Model - Python implementation using ViennaPS.
    Replicates oxidationFront.py functionality.
    """
    def __init__(self, config_file):
        self.params = vps.readConfigFile(config_file)
        
        self.dim = int(self.params.get("dimensions", 3))
        vps.setDimension(self.dim)
        vps.setNumThreads(int(self.params.get('numThreads', 1)))
        
        # 1. Geometry
        ls_list, mat_map, _, _, grid_delta = geometryPS.get_geometry_domains(self.params)
        self.grid_delta = grid_delta
        
        # 2. DenseCellSet
        print("Building DenseCellSet...")
        # ViennaPS exposes DenseCellSet via vps.d2 or vps.d3 based on setDimension
        self.cell_set = vps.DenseCellSet()
        
        depth = self.params['substrateHeight'] + self.params.get('ambientHeight', grid_delta)
        self.cell_set.setCellSetPosition(True)
        self.cell_set.setCoverMaterial(MAT_AMBIENT) # Ensure gas is MAT_AMBIENT
        self.cell_set.fromLevelSets(ls_list, mat_map, depth)
        self.cell_set.buildNeighborhood()
        self.num_cells = self.cell_set.getNumberOfCells()
        
        # 3. Data Fields
        self.cell_set.addScalarData("oxidant", 0.0)
        self.cell_set.addScalarData("oxideFraction", 0.0)
        
        # Python State
        self.materials = np.array(self.cell_set.getScalarData("Material"), dtype=int)
        self.oxidant = np.zeros(self.num_cells)
        self.oxide_fraction = np.zeros(self.num_cells)
        self.e_field_1d = np.zeros(self.num_cells)

        # Pre-calculate neighbor map
        self._build_neighbor_map()
        
        # 4. Initialize oxidant in ambient cells
        is_ambient = self.materials == MAT_AMBIENT
        self.oxidant[is_ambient] = self.params['ambientOxidant']
        
        # 5. Pre-compute E-Field mapping
        self._precompute_efield_indices()

        # 6. Build Topology
        print("Building Sparse Topology...")
        self._build_topology()

        # 7. Sync initial state and write
        self._sync_data()
        self.cell_set.writeVTU("oxidation_initial.vtu")

    def _build_neighbor_map(self):
        print("Building neighbor map for vectorization...")
        self.max_neighbors = 2 * self.dim
        self.neighbor_map = np.full((self.num_cells, self.max_neighbors), -1, dtype=np.int32)

        for idx in range(self.num_cells):
            nbs = self.cell_set.getNeighbors(idx)
            count = min(len(nbs), self.max_neighbors)
            if count > 0:
                self.neighbor_map[idx, :count] = nbs[:count]

    def _sync_data(self):
        self.cell_set.setScalarData("oxidant", self.oxidant)
        self.cell_set.setScalarData("oxideFraction", self.oxide_fraction)
        self.cell_set.setScalarData("Material", self.materials.astype(float))

    def _precompute_efield_indices(self):
        csv_dx = 1.0
        csv_extent = 150.0
        nx = int(csv_extent / csv_dx)

        centers = np.array([self.cell_set.getCellCenter(i) for i in range(self.num_cells)])
        gen_x = centers[:, 0] + (csv_extent / 2.0)
        ix = np.clip((gen_x / csv_dx).astype(int), 0, nx - 1)

        if self.dim == 2:
            self.efield_grid_indices = ix
        else:
            gen_y = centers[:, 1] + (csv_extent / 2.0)
            iy = np.clip((gen_y / csv_dx).astype(int), 0, nx - 1)
            self.efield_grid_indices = iy * nx + ix

    def _load_efield_data(self, time):
        filename = self.params['EfieldFile']
        if not os.path.exists(filename):
            print(f"[ERROR] {filename} not found.")
            sys.exit(1)

        try:
            raw_data = np.genfromtxt(filename, delimiter=',')
        except Exception as e:
            print(f"[ERROR] Reading E-field: {e}")
            sys.exit(1)

        if len(raw_data.shape) == 1:
            return np.array([raw_data[-1]])

        if self.dim == 2:
            return raw_data[:, 1] if raw_data.shape[1] >= 2 else raw_data[:, -1]
        else:
            return raw_data[:, 2] if raw_data.shape[1] >= 3 else raw_data[:, -1]

    def _update_electric_field(self, time):
        e_field_values = self._load_efield_data(time)
        valid_mask = (self.efield_grid_indices >= 0) & (self.efield_grid_indices < len(e_field_values))
        self.e_field_1d[:] = 0.0
        self.e_field_1d[valid_mask] = e_field_values[self.efield_grid_indices[valid_mask]]

    def _build_topology(self):
        is_mat_active = (self.materials == MAT_SUBSTRATE) | (self.materials == MAT_OXIDE)
        cond_oxide = self.oxide_fraction > 1e-6

        valid_nb = self.neighbor_map >= 0
        safe_nbs = np.where(valid_nb, self.neighbor_map, 0)
        nb_mats = self.materials[safe_nbs]
        nb_is_ambient = (nb_mats == MAT_AMBIENT) & valid_nb
        touching_ambient = np.any(nb_is_ambient, axis=1)

        is_core_active = is_mat_active & (cond_oxide | touching_ambient)

        nb_is_core = is_core_active[safe_nbs] & valid_nb
        touching_core = np.any(nb_is_core, axis=1)

        is_active = is_mat_active & (is_core_active | touching_core)

        self.active_indices = np.where(is_active)[0]
        self.is_active_cell = is_active
        self.n_dof = len(self.active_indices)
        print(f"Active cells: {self.n_dof} / {self.num_cells}")

        self._build_diffusion_structure()

    def _build_diffusion_structure(self):
        active_idxs = self.active_indices
        self.ambient_neighbor_count = np.zeros(self.num_cells, dtype=np.int32)

        if len(active_idxs) == 0:
            self.diff_rows = np.array([], dtype=np.int64)
            self.diff_cols = np.array([], dtype=np.int64)
            return

        nbs = self.neighbor_map[active_idxs]
        valid_nb = nbs >= 0
        safe_nbs = np.where(valid_nb, nbs, 0)
        mat_nbs = self.materials[safe_nbs]

        rows_list = []
        cols_list = []

        for k in range(self.max_neighbors):
            valid = valid_nb[:, k]
            mat_n = mat_nbs[:, k]
            idx_n = safe_nbs[:, k]

            is_amb = (mat_n == MAT_AMBIENT) & valid
            np.add.at(self.ambient_neighbor_count, active_idxs[is_amb], 1)

            is_active_nb = (self.is_active_cell[idx_n] & valid & (mat_n != MAT_AMBIENT) & (mat_n != MAT_MASK))
            if np.any(is_active_nb):
                rows_list.append(active_idxs[is_active_nb])
                cols_list.append(idx_n[is_active_nb])

        self.diff_rows = np.concatenate(rows_list) if rows_list else np.array([], dtype=np.int64)
        self.diff_cols = np.concatenate(cols_list) if cols_list else np.array([], dtype=np.int64)

    def solve_step(self, dt):
        dx = self.grid_delta
        dtdx2 = dt / (dx * dx)
        ambient_oxidant = self.params['ambientOxidant']
        D_ox = self.params['oxidantDiffusivity']

        k_base = self.params['reactionRateConstant']
        alpha = self.params['eFieldInfluence']
        E_mag = np.abs(self.e_field_1d)
        available_si = np.maximum(0.0, 1.0 - self.oxide_fraction)
        reaction_rates = k_base * (1.0 + alpha * E_mag) * available_si

        active_idxs = self.active_indices
        if len(active_idxs) > 0:
            C_old = self.oxidant[active_idxs]
            k = reaction_rates[active_idxs]
            n = self.num_cells

            if len(self.diff_rows) > 0:
                D_eff = 0.5 * D_ox * (self.oxide_fraction[self.diff_rows] + self.oxide_fraction[self.diff_cols])
                diag = np.zeros(n)
                np.add.at(diag, self.diff_rows, -D_eff)
                diag[active_idxs] -= self.ambient_neighbor_count[active_idxs] * D_ox

                all_rows = np.concatenate([self.diff_rows, active_idxs])
                all_cols = np.concatenate([self.diff_cols, active_idxs])
                all_vals = np.concatenate([D_eff, diag[active_idxs]])

                A = coo_matrix((all_vals, (all_rows, all_cols)), shape=(n, n)).tocsr()
                diff_term = A.dot(self.oxidant)[active_idxs]
            else:
                diff_term = -self.ambient_neighbor_count[active_idxs] * D_ox * C_old

            diff_term += self.ambient_neighbor_count[active_idxs] * D_ox * ambient_oxidant
            self.oxidant[active_idxs] = np.maximum(0.0, C_old + dtdx2 * diff_term - dt * k * C_old)

        mask_update = ((self.materials == MAT_SUBSTRATE) | (self.materials == MAT_OXIDE)) & (self.oxidant > 1e-12)
        idxs_update = np.where(mask_update)[0]
        
        if len(idxs_update) > 0:
            C = self.oxidant[idxs_update]
            k = reaction_rates[idxs_update]
            d_frac = k * C * dt
            
            self.oxide_fraction[idxs_update] += d_frac
            self.oxide_fraction[idxs_update] = np.minimum(1.0, self.oxide_fraction[idxs_update])
            
            new_oxides = (self.oxide_fraction[idxs_update] > 0.5) & (self.materials[idxs_update] == MAT_SUBSTRATE)
            if np.any(new_oxides):
                self.materials[idxs_update[new_oxides]] = MAT_OXIDE
                return True
        return False

    def run(self):
        duration = self.params['duration']
        D_ox = self.params['oxidantDiffusivity']
        stability_factor = 2 * self.dim
        dt_base = (self.grid_delta**2 / (D_ox * stability_factor)) * self.params['timeStabilityFactor']

        time = 0.0
        step = 0
        topology_rebuild_interval = int(self.params.get('topologyRebuildInterval', 5))
        need_topology_rebuild = True
        print(f"Starting simulation. Duration: {duration}, initial dt: {dt_base:.4f}")
        
        while time < duration:
            if need_topology_rebuild or step % topology_rebuild_interval == 0:
                self._build_topology()
                need_topology_rebuild = False

            self._update_electric_field(time)

            if len(self.active_indices) > 0:
                k_base = self.params['reactionRateConstant']
                alpha = self.params['eFieldInfluence']
                E_mag = np.abs(self.e_field_1d[self.active_indices])
                avail_si = np.maximum(0.0, 1.0 - self.oxide_fraction[self.active_indices])
                rates = k_base * (1.0 + alpha * E_mag) * avail_si
                max_k = np.max(rates)
            else:
                max_k = 0.0

            dt_diff = self.grid_delta**2 / (D_ox * stability_factor)
            dt_react = (1.0 / max_k) if max_k > 1e-12 else duration
            dt = min(dt_diff, dt_react) * self.params['timeStabilityFactor']

            if time + dt > duration:
                dt = duration - time

            material_changed = self.solve_step(dt)
            if material_changed:
                need_topology_rebuild = True
            time += dt
            step += 1
            
            if step % 10 == 0:
                self._sync_data()
                print(f"Step {step}, Time: {time:.4f}, dt: {dt:.4f}, max_k: {max_k:.4f}")
                self.cell_set.writeVTU(f"oxidation_step_{step}.vtu")
                
        self._sync_data()
        self.cell_set.writeVTU("oxidation_final.vtu")
        print("Simulation Done.")


if __name__ == "__main__":
    config_file = "config.txt"
    if len(sys.argv) > 1:
        config_file = sys.argv[1]
    
    sim = OxidationSimulation(config_file)
    sim.run()

File: examples/oxidationFrontPS/oxidationFrontPS.cpp
#include <algorithm>
#include <array>
#include <cmath>
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>
#include <unordered_map>

#include <lsAdvect.hpp>
#include <psDomain.hpp>
#include <process/psProcess.hpp>
#include <process/psTranslationField.hpp>

// #include <psUtils.hpp>

// ViennaCS includes are available since ViennaPS links against it
#include <csDenseCellSet.hpp>

#include "geometryPS.hpp"

namespace ps = viennaps;
namespace cs = viennacs;
namespace ls = viennals;

using NumericType = double;
constexpr int D = 3;

// Material IDs
constexpr ps::Material MAT_SUBSTRATE = ps::Material::Si;
constexpr ps::Material MAT_OXIDE = ps::Material::SiO2; // We will use this for the oxide layer
constexpr ps::Material MAT_MASK = ps::Material::Mask;
constexpr ps::Material MAT_AMBIENT = ps::Material::GAS;
constexpr ps::Material MAT_VOID = ps::Material::Undefined;

class OxidationSimulation {
public:
  OxidationSimulation(const std::string &configFile) {
    params.readConfigFile(configFile);
    omp_set_num_threads(params.get<int>("numThreads"));
    
    ps::Logger::setLogLevel(ps::LogLevel::DEBUG);

    // Cache parameters
    k_base = params.get("reactionRateConstant");
    alpha = params.get("eFieldInfluence");
    D_ox = params.get("oxidantDiffusivity");
    ambientOxidant = params.get("ambientOxidant");
    dx = params.get("gridDelta");
    duration = params.get("duration");
    timeStabilityFactor = params.get("timeStabilityFactor");
    try {
      topologyRebuildInterval = params.get<int>("topologyRebuildInterval");
    } catch (...) {
      topologyRebuildInterval = 5;
    }

    // Initialize Domain and Geometry
    initGeometry();
  }

  void apply() {
    NumericType time = 0.;
    int step = 0;
    bool materialChanged = true;

    // Initial CellSet generation
    std::cout << "Generating initial CellSet..." << std::endl;
    generateCellSet();
    precomputeEFieldIndices();
    initOxidationFields();
    
    // Write initial state
    domain->getCellSet()->writeVTU("oxidation_initial.vtu");

    while (time < duration) {
      std::cout << "--- Step " << step << " (Time: " << time << ") ---" << std::endl;
      
      if (step % topologyRebuildInterval == 0 || materialChanged) {
        updateActiveCells();
        materialChanged = false;
      }

      loadElectricField();
      // Calculate max reaction rate for time step control
      calculateReactionRates();
      NumericType max_k = calculateMaxReactionRate();

      // 1. Determine Time Step
      NumericType dt_diff = dx * dx / (D_ox * 2 * D);
      NumericType dt_react = (max_k > 1e-12) ? (1.0 / max_k) : duration;
      // Use dt_react as limit for interface advection stability
      NumericType dt = std::min(dt_diff, dt_react);// * timeStabilityFactor;

      if (time + dt > duration) dt = duration - time;
      if (step == 0) std::cout << "Initial dt: " << dt << std::endl;

      // 2. Solve Diffusion (Explicit)
      std::cout << "Solving diffusion (Explicit)..." << std::endl;
      solveDiffusionExplicit(dt);

      // 3. Update Oxide Fraction (ViennaCS logic)
      // This evolves the materials in the CellSet
      if (updateOxideFraction(dt)) {
        materialChanged = true;
      }

      // 4. Advect Interface (ViennaPS feature)
      // Calculate velocity v = -k * C_ox at the interface
      // We store this velocity on the CellSet, ViennaPS will map it to the Level Set
      calculateVelocity();
      std::cout << "Advecting interface..." << std::endl;

      {
        auto velocityField = ps::SmartPointer<CustomVelocityField>::New(cellSet, "velocity");
        auto translationField = ps::SmartPointer<ps::TranslationField<NumericType, D>>::New(
            velocityField, domain->getMaterialMap(), 0);
        ls::Advect<NumericType, D> advectionKernel;
        advectionKernel.insertNextLevelSet(domain->getLevelSets()[0]);
        advectionKernel.setVelocityField(translationField);
        advectionKernel.setAdvectionTime(dt);
        advectionKernel.apply();
      }

      time += dt;
      step++;

      if (step % 10 == 0) {
        std::cout << "Step " << step << " Time: " << time << " dt: " << dt << std::endl;
        cellSet->writeVTU("oxidation_step_" + std::to_string(step) + ".vtu");

        auto mesh = ls::SmartPointer<ls::Mesh<NumericType>>::New();
        ls::ToSurfaceMesh<NumericType, D>(domain->getLevelSets()[0], mesh).apply();
        ls::VTKWriter<NumericType>(mesh, "oxidation_ls_step_" + std::to_string(step) + ".vtp").apply();
      }
    }

    cellSet->writeVTU("oxidation_final.vtu");
    std::cout << "Simulation Done." << std::endl;
  }

private:
  ps::util::Parameters params;
  ps::SmartPointer<ps::Domain<NumericType, D>> domain;
  ps::SmartPointer<cs::DenseCellSet<NumericType, D>> cellSet;

  class CustomVelocityField : public ps::VelocityField<NumericType, D> {
    ps::SmartPointer<cs::DenseCellSet<NumericType, D>> cellSet;
    std::string velocityName;
    std::vector<NumericType> *velocityData = nullptr;

  public:
    CustomVelocityField(ps::SmartPointer<cs::DenseCellSet<NumericType, D>> cs,
                        std::string name)
        : cellSet(cs), velocityName(name) {
        velocityData = cellSet->getScalarData(velocityName);
    }

    NumericType getScalarVelocity(const viennacore::Vec3D<NumericType> &coord, int,
                                  const viennacore::Vec3D<NumericType> &,
                                  unsigned long) override {
      int idx = cellSet->getIndex(coord);
      if (idx >= 0) {
        if (velocityData && idx < velocityData->size())
          return (*velocityData)[idx];
      }
      return 0.0;
    }

    viennacore::Vec3D<NumericType> getVectorVelocity(const viennacore::Vec3D<NumericType> &, int,
                                                 const viennacore::Vec3D<NumericType> &,
                                                 unsigned long) override {
      return {0., 0., 0.};
    }
  };

  NumericType k_base, alpha, D_ox, ambientOxidant, dx, duration, timeStabilityFactor;
  std::vector<NumericType> electricField1D;
  std::vector<int> efieldIndices;
  std::vector<NumericType> reactionRates;
  std::vector<NumericType> nextOxidant;
  std::vector<int> activeCells;
  std::vector<char> isActiveCell;
  int topologyRebuildInterval;

  void initGeometry() {
    // Create Domain
    domain = ps::SmartPointer<ps::Domain<NumericType, D>>::New(
        params.get("gridDelta"), params.get("xExtent"), params.get("yExtent"));

    // Use geometryPS.hpp to match ViennaCS example
    auto matMap = ls::SmartPointer<ls::MaterialMap>::New();
    std::vector<int> materials;
    auto levelSets = geometry::makeStructure<NumericType, D>(
        params, matMap, materials, static_cast<int>(MAT_SUBSTRATE),
        static_cast<int>(MAT_MASK), static_cast<int>(MAT_AMBIENT));

    for (size_t i = 0; i < levelSets.size(); ++i) {
      domain->insertNextLevelSetAsMaterial(levelSets[i],
                                           static_cast<ps::Material>(materials[i]));
    }
  }

  void generateCellSet() {
    NumericType depth = params.get("substrateHeight") + params.get("maskHeight") + params.get("ambientHeight");
    // Generate CellSet. "true" indicates we want cells above the surface (gas), 
    // but here we want the whole domain to solve diffusion.
    domain->generateCellSet(depth, MAT_VOID, true);
    cellSet = std::dynamic_pointer_cast<cs::DenseCellSet<NumericType, D>>(domain->getCellSet());
    cellSet->buildNeighborhood();
  }

  void loadElectricField() {
    electricField1D.clear();
    std::string csvFileName = params.get<std::string>("EfieldFile");

    std::ifstream in(csvFileName);
    if (!in.good()) {
        std::cerr << "Error: cannot open file " << csvFileName << std::endl;
        return;
    }

    std::string line;
    while (std::getline(in, line)) {
        if (line.empty() || line[0] == '#' || line[0] == '%') continue;
        std::stringstream ss(line);
        std::string segment;
        std::vector<NumericType> row;
        while (std::getline(ss, segment, ',')) {
            try { row.push_back(std::stod(segment)); } catch (...) {}
        }
        
        // Match oxidationFront.cpp parsing logic
        if (D == 2 && row.size() >= 2) {
            electricField1D.push_back(row[1]);
        } else if (D == 3 && row.size() >= 3) {
            electricField1D.push_back(row[2]);
        } else if (!row.empty()) {
            electricField1D.push_back(row.back());
        }
    }
  }

  void precomputeEFieldIndices() {
    efieldIndices.resize(cellSet->getNumberOfCells());
    // Match oxidationFront.cpp mapping constants
    const NumericType csvDx = 1.0;
    const NumericType csvExtent = 150.0;
    const int nx = static_cast<int>(csvExtent / csvDx);

    #pragma omp parallel for
    for (int i = 0; i < cellSet->getNumberOfCells(); ++i) {
        auto center = cellSet->getCellCenter(i);
        // Map coordinates: Sim X [-x/2, x/2] -> Gen X [0, x] using csvExtent
        NumericType gen_x = center[0] + (csvExtent / 2.0);

        int ix = static_cast<int>(gen_x / csvDx);
        int idx = -1;

        if constexpr (D == 2) {
            idx = ix;
        } else {
            NumericType gen_y = center[1] + (csvExtent / 2.0);
            int iy = static_cast<int>(gen_y / csvDx);
            if (iy >= 0 && iy < nx) idx = iy * nx + ix;
        }

        efieldIndices[i] = idx;
    }
  }

  void initOxidationFields() {
    auto oxidant = cellSet->addScalarData("oxidant", 0.0);
    auto oxideFraction = cellSet->addScalarData("oxideFraction", 0.0);
    auto materials = cellSet->getScalarData("Material");
    
    if (!materials) {
        std::cerr << "Error: Material field missing in initOxidationFields!" << std::endl;
        return;
    }

    // Initialize oxideFraction based on Material ID to drive diffusion
    // In ViennaPS, the geometry is explicit, so we set fraction=1 in Oxide, 0 in Substrate
    #pragma omp parallel for
    for (int i = 0; i < cellSet->getNumberOfCells(); ++i) {
      int mat = static_cast<int>((*materials)[i]);
      if (mat == static_cast<int>(MAT_AMBIENT)) {
        (*oxidant)[i] = ambientOxidant;
        (*oxideFraction)[i] = 0.0; // Gas
      } else if (mat == static_cast<int>(MAT_OXIDE)) {
        (*oxideFraction)[i] = 1.0;
      } else {
        (*oxideFraction)[i] = 0.0;
      }
    }
  }

  NumericType getReactionRate(NumericType E_mag, NumericType oxideFrac) {
    NumericType availableSi = std::max(0.0, 1.0 - oxideFrac);
    return k_base * (1.0 + alpha * std::abs(E_mag)) * availableSi;
  }

  NumericType getDiffusivity(int material, NumericType oxideFrac) {
    if (material == static_cast<int>(MAT_MASK)) return 0.0;
    if (material == static_cast<int>(MAT_AMBIENT)) return 0.0; // Dirichlet BC handled separately
    return D_ox * oxideFrac;
  }

  void calculateReactionRates() {
      if (reactionRates.size() != cellSet->getNumberOfCells()) {
          reactionRates.resize(cellSet->getNumberOfCells());
      }
      auto oxideFractions = cellSet->getScalarData("oxideFraction");
      auto materials = cellSet->getScalarData("Material");

      #pragma omp parallel for
      for (int i = 0; i < cellSet->getNumberOfCells(); ++i) {
          int mat = static_cast<int>((*materials)[i]);
          if (mat == static_cast<int>(MAT_SUBSTRATE) || mat == static_cast<int>(MAT_OXIDE)) {
              NumericType E = 0.0;
              int idx = efieldIndices[i];
              if (idx >= 0 && idx < electricField1D.size()) E = electricField1D[idx];
              reactionRates[i] = getReactionRate(E, (*oxideFractions)[i]);
          } else {
              reactionRates[i] = 0.0;
          }
      }
  }

  void updateActiveCells() {
    auto materials = cellSet->getScalarData("Material");
    auto oxideFractions = cellSet->getScalarData("oxideFraction");

    activeCells.clear();
    isActiveCell.assign(cellSet->getNumberOfCells(), 0);

    // Track which cells are "Core Active" (Oxide or Source)
    std::vector<char> isCoreActive(cellSet->getNumberOfCells(), 0);

    #pragma omp parallel for
    for (int i = 0; i < cellSet->getNumberOfCells(); ++i) {
      int mat = static_cast<int>((*materials)[i]);
      if (mat == static_cast<int>(MAT_SUBSTRATE) ||
          mat == static_cast<int>(MAT_OXIDE)) {
        bool active = (*oxideFractions)[i] > 1e-6;
        if (!active) {
          // Check if touching ambient (Source term)
          auto neighbors = cellSet->getNeighbors(i);
          for (auto n : neighbors) {
            if (n >= 0) {
                int mat_n = static_cast<int>((*materials)[n]);
                if (mat_n == static_cast<int>(MAT_AMBIENT)) {
                    active = true;
                    break;
                }
            }
          }
        }
        isCoreActive[i] = active;
      }
    }

    // Expand to neighbors to allow front propagation
    #pragma omp parallel for
    for (int i = 0; i < cellSet->getNumberOfCells(); ++i) {
      int mat = static_cast<int>((*materials)[i]);
      if (mat == static_cast<int>(MAT_SUBSTRATE) ||
          mat == static_cast<int>(MAT_OXIDE)) {
        // Solve if Core Active OR neighbor of Core Active (Narrow Band)
        bool shouldSolve = isCoreActive[i];
        if (!shouldSolve) {
          auto neighbors = cellSet->getNeighbors(i);
          for (auto n : neighbors) {
            if (n >= 0 && isCoreActive[n]) {
              shouldSolve = true;
              break;
            }
          }
        }

        if (shouldSolve) {
          isActiveCell[i] = 1;
        }
      }
    }

    for (int i = 0; i < cellSet->getNumberOfCells(); ++i) {
      if (isActiveCell[i]) activeCells.push_back(i);
    }
  }

  void solveDiffusionExplicit(NumericType dt) {
    auto oxidant = cellSet->getScalarData("oxidant");
    auto materials = cellSet->getScalarData("Material");
    auto oxideFractions = cellSet->getScalarData("oxideFraction");

    if (nextOxidant.size() != cellSet->getNumberOfCells()) {
      nextOxidant.resize(cellSet->getNumberOfCells());
    }

    NumericType dtdx2 = dt / (dx * dx);

    #pragma omp parallel for
    for (int idx = 0; idx < static_cast<int>(activeCells.size()); ++idx) {
        int i = activeCells[idx];
        int mat = static_cast<int>((*materials)[i]);

        NumericType C_old = (*oxidant)[i];
        NumericType k = reactionRates[i];
        NumericType diffusion_term = 0.0;
        NumericType D_center = getDiffusivity(mat, (*oxideFractions)[i]);

        auto neighbors = cellSet->getNeighbors(i);
        for (int n : neighbors) {
          if (n < 0) continue;
          int mat_n = static_cast<int>((*materials)[n]);
          
          if (mat_n == static_cast<int>(MAT_MASK) || mat_n == static_cast<int>(MAT_VOID)) continue;

          if (mat_n == static_cast<int>(MAT_AMBIENT)) {
            // Dirichlet BC
            diffusion_term += D_ox * (ambientOxidant - C_old);
          } else if (isActiveCell[n]) {
             // Neighbor is Substrate or Oxide and Active
             NumericType D_neighbor = getDiffusivity(mat_n, (*oxideFractions)[n]);
             NumericType D_eff = 0.5 * (D_center + D_neighbor);
             diffusion_term += D_eff * ((*oxidant)[n] - C_old);
          }
        }
        
        nextOxidant[i] = C_old + dtdx2 * diffusion_term - dt * k * C_old;
    }

    // Update
    #pragma omp parallel for
    for (int idx = 0; idx < static_cast<int>(activeCells.size()); ++idx) {
        int i = activeCells[idx];
        (*oxidant)[i] = nextOxidant[i];
    }
  }

  bool updateOxideFraction(NumericType dt) {
    auto oxidant = cellSet->getScalarData("oxidant");
    auto oxideFractions = cellSet->getScalarData("oxideFraction");
    auto materials = cellSet->getScalarData("Material");
    int materialChanged = 0;

    #pragma omp parallel for reduction(+:materialChanged)
    for (int i = 0; i < cellSet->getNumberOfCells(); ++i) {
      int mat = static_cast<int>((*materials)[i]);
      if (mat == static_cast<int>(MAT_SUBSTRATE) || mat == static_cast<int>(MAT_OXIDE)) {
        NumericType C = (*oxidant)[i];
        if (C > 1e-12) {
           NumericType k = reactionRates[i];
           NumericType dFrac = k * C * dt;
           (*oxideFractions)[i] += dFrac;
           if ((*oxideFractions)[i] > 1.0) (*oxideFractions)[i] = 1.0;
           
           if ((*oxideFractions)[i] > 0.5 && mat == static_cast<int>(MAT_SUBSTRATE)) {
               (*materials)[i] = static_cast<int>(MAT_OXIDE);
               materialChanged = 1;
           }
        }
      }
    }
    return materialChanged > 0;
  }

  void calculateVelocity() {
    auto velocity = cellSet->addScalarData("velocity", 0.0);
    auto oxidant = cellSet->getScalarData("oxidant");
    auto materials = cellSet->getScalarData("Material");

    if (!oxidant || !materials || !velocity) return;

    #pragma omp parallel for
    for (int i = 0; i < cellSet->getNumberOfCells(); ++i) {
      int mat = static_cast<int>((*materials)[i]);
      // Calculate velocity in both Substrate and Oxide to ensure robust tracking
      if (mat == static_cast<int>(MAT_SUBSTRATE) || mat == static_cast<int>(MAT_OXIDE)) {
          NumericType E = 0.0;
          int idx = efieldIndices[i];
          if (idx >= 0 && idx < electricField1D.size()) E = electricField1D[idx];
          
          // Use potential rate (assuming full Si availability) for LS velocity.
          // This prevents the LS from stopping if it lags into the fully oxidized region.
          NumericType k_potential = k_base * (1.0 + alpha * std::abs(E));
          
          // v = -k * C * dx (Negative because the interface moves into the substrate)
          (*velocity)[i] = -k_potential * (*oxidant)[i] * dx;
      } else {
          (*velocity)[i] = 0.0;
      }
    }

    // Extend velocity into the substrate to ensure the Level Set interface (which lies between
    // active and inactive cells) sees the peak reaction velocity, preventing lag.
    std::vector<NumericType> originalVelocity = *velocity;
    #pragma omp parallel for
    for (int i = 0; i < cellSet->getNumberOfCells(); ++i) {
        int mat = static_cast<int>((*materials)[i]);
        if (mat == static_cast<int>(MAT_SUBSTRATE)) {
            NumericType max_speed = std::abs(originalVelocity[i]);
            for (int n : cellSet->getNeighbors(i)) { // Check all neighbors
                if (n >= 0) {
                    max_speed = std::max(max_speed, std::abs(originalVelocity[n]));
                }
            }
            (*velocity)[i] = -max_speed;
        }
    }
  }

  NumericType calculateMaxReactionRate() {
    NumericType max_k = 0.0;
    
    #pragma omp parallel for reduction(max:max_k)
    for (int i = 0; i < reactionRates.size(); ++i) {
        if (reactionRates[i] > max_k) max_k = reactionRates[i];
    }
    return max_k;
  }
};

int main(int argc, char **argv) {
  std::string configFile = "config.txt";
  if (argc > 1) configFile = argv[1];
  
  OxidationSimulation sim(configFile);
  sim.apply();
  return 0;
}